fixed(toString): discard undefined parameters from route params

Eldow opened the pull request:
This is the pull request refering to #36 

It works fine with undefined parameters, but I tried the same test with null parameters and another problem showed up as they were parsed back as none.

``` typescript
 it('should be reentrant', () => {
    const dummy = lit('x').then(
      query(t.interface({ a: t.union([t.null, t.string]) }))
    );
    assert.deepEqual(
      dummy.parser.run(Route.parse(format(dummy.formatter, { a: null }))),
      some([{ a: null }, Route.empty])
    ); // errors as parsed as none
  });
```
@gcanti What is your opinion, is this the intended behaviour  ?

gcanti commented:
@Eldow my opinion is that `null` should be removed from the definition of `Query`

```ts
export interface Query {
  [key: string]: string | Array<string> | undefined
}
```

since is not clear what's the serializated form of a `null` (should be an explicit decision by the user).

@sledorze ?

Eldow commented:
I upgraded fp-ts to 1.10 as Record wan't part of 1.0

gcanti commented:
> is not clear what's the serialized form of a null

@sledorze ok, so you are proposing to assume the same serialization for both `null` and `undefined`, this way any runtime type which is using `fromNullable` internally (like `createOptionFromNullable`) is automatically handled, right?

@Eldow I think we could use `filter` instead of `filterMap` in order to prevent the creation of `Option`s

```ts
import { filter } from 'fp-ts/lib/Record'

...

const qs = stringify(filter(this.query, value => value != null))
```

Also, the CI is failing because you must upgrade `fp-ts` through `npm install`

sledorze commented:
@gcanti yes, I'm proposing exactly that.

sledorze commented:
As far as null is concerned, I think that the relevant test would exhibit the fact that a nullable property of an object which is not present in the url would translate into an entry in the parsed response with a 'null' value.

e.g.:
```typescript
  it('toString discards null parameters', () => {
    const stringOrNull = t.union([t.null, t.string])
    const dummy = lit('x').then(query(t.interface({ a: stringOrNull, b: stringOrNull })))
    assert.deepEqual(
      dummy.parser.run(Route.parse(format(dummy.formatter, { a: null, b: 'evidence' }))),
      some([{ a: null, b: 'evidence' }, Route.empty])
    )
  })
```
Achieving this would requiers to preemptively select the properties which can validates nulls and then when a new object to decode arrives, insert some nulls when an entry is missing.

For that purpose we need to export that from `io-ts`:

https://github.com/gcanti/io-ts/blob/master/src/index.ts#L1294

@gcanti @Eldow what do you think?

Eldow commented:
I've added in the PR the sample code related to the modification proposed by @sledorze to give you an overview of what we're willing to implement. We've added the function that would need to be exported from io-ts in an external file for the example purpose.

Tell me what do you think of it @gcanti

gcanti commented:
Looks too much complicated and with little gain over a user-defined `NullableString` runtime type

```ts
const nullable = <A>(type: t.Type<A, A>, name: string): t.Type<A | null, A | null> => {
  return new t.Type<A | null>(
    name,
    (m): m is A | null => m === null || type.is(m),
    (i, c) => (i == null ? t.success(null) : type.validate(i, c)),
    t.identity
  )
}

const NullableString = nullable(t.string, 'NullableString')

const dummy = lit('x').then(query(t.interface({ a: NullableString, b: NullableString })))

assert.deepEqual(
  dummy.parser.run(Route.parse(format(dummy.formatter, { a: null, b: 'evidence' }))),
  some([{ a: null, b: 'evidence' }, Route.empty])
) // => ok
```

sledorze commented:
As far as type-safety is concerned, I think the implementation proposed by @Eldow is more reliable (but more restrictive by preventing custom types usage).

Looks like a choice between simple and correct.

gcanti commented:
@sledorze yeah, that's the consequence of assuming

> the same serialization for both `null` and `undefined`

So either we remove `null` from

```ts
export interface Query {
  [key: string]: string | Array<string> | undefined
}
```

and force users to convert `null`s to `undefined`s (back and forth) explicitly..

..or we keep `null`s in `Query`, we discard them automatically, and then users must choose their runtime types wisely

> As far as type-safety is concerned

Note: type-safety (meaning no runtime errors) is ok in both cases, in the second case though if you choose the wrong runtime type it won't work as you may expect

sledorze commented:
Removing null from Query values Type means `createOptionFromNullable` from `io-ts-types` can no more be used.

https://github.com/gcanti/io-ts-types/blob/master/src/fp-ts/createOptionFromNullable.ts#L20

Honestly, at this point; it either 'we have covered it in the lib once and forall' or 'you users must know that particular restriction: ...'.

gcanti commented:
@sledorze I agree, that's why I prefer the first case

> Removing null from Query values Type means createOptionFromNullable from io-ts-types can no more be used

It's just a single `mapOutput` away though

```ts
const StringOption = mapOutput(createOptionFromNullable(t.string), a => (a === null ? undefined : a))
```

Again in my opinion the simplest and neatest solution is 

1) removing `null` from `Query`

```ts
export interface Query {
  [key: string]: string | Array<string> | undefined
}
```

2) filter out `undefined`s from `this.query` in the `Route.prototype.toString()` method (as per this PR)

---

After doing that, you'll have several ways to encode optional query params without any drawbacks

```ts
import { none, some } from 'fp-ts/lib/Option'
import * as t from 'io-ts'
import { createOptionFromNullable, mapOutput } from 'io-ts-types'
import { format, Match, query, Route } from '../src'
import assert = require('assert')

const StringOption = mapOutput(createOptionFromNullable(t.string), a => (a === null ? undefined : a))

const q1 = query(t.partial({ foo: t.string }))
const q2 = query(t.type({ foo: t.union([t.undefined, t.string]) }))
const q3 = query(t.type({ foo: StringOption }))

const testRoundtrip = <A extends object>(q: Match<A>, a: A, expected: A): void => {
  assert.deepEqual(q.parser.run(Route.parse(format(q.formatter, a))), some([expected, Route.empty]))
}

testRoundtrip(q1, {}, {})
testRoundtrip(q1, { foo: undefined }, {})
testRoundtrip(q2, { foo: undefined }, { foo: undefined })
testRoundtrip(q3, { foo: none }, { foo: none })
testRoundtrip(q3, { foo: some('bar') }, { foo: some('bar') })
```

sledorze commented:
My thinking:

@gcanti your solution is fine provided that one can use an additional combinator on top of its type to do the transformation @Eldow proposed to integrate.

Two things in my mind:

1) Can we export `getProps` from `io-ts` so that if someone needs to implement such feature, he can do it.
2) That combinator (or a generalisation of it) may be added to `io-ts-types`.

Then all possibilities remain available, `fp-ts-routing` stays simple and we can explain the pattern in the Readme.

@Eldow what do you think?

sledorze commented:
Most of the types we're using come from codegen from IDLs (like swagger).
So if I'm looking for a general solution that handles both `undefined` and `null` consistently, without requiring additional manual modification or having to implement different logic between query params and the rest of the codegen.
That can be done, but it's another thing that adds to complexity.
However, the solution I exposed in my last comment is satisfying.
To be honest I have liked it to be handled natively by `fp-ts-routing` (`null` and `undefined` begin some primitives of the platform we're using) but I can have a generic solution with just and additional combinator, so that's a workable solution.

gcanti commented:
Ok, anyway we can do one step at the time:

1) let's remove `null` from `Query` and filter out `undefined`s from `this.query` in the `Route.prototype.toString()` method..

> However, the solution I exposed in my last comment is satisfying

2) ..and let's postpone the problem of handling `null`s natively until you tried your solution on the field

sledorze commented:
I agree, @Eldow does that work for you? (We need a updated PR)

Eldow commented:
I updated the PR following your recommendations, let me know if that works for you @sledorze @gcanti

sledorze commented on src/index.ts:
```diff
@@ -35,7 +34,7 @@ export class Route {
   }
   toString(encode: boolean = true): string {
-    const nonNullQuery = records.filter(this.query, part => part !== undefined && part !== null)
+    const nonNullQuery = records.filter(this.query, part => part !== undefined)
```
@Eldow the naming is now wrong (nonUndefinedQuery)?

Eldow replied:
oh that's right

gcanti commented:
Thanks @Eldow
